pkg/resource/restconfig_factory.go

```diff
@@ -2,30 +2,115 @@ package resource
+// eventually the inner NewDiscoveryRESTMapper() function is invoked, which is very expensive
+// and allocates a huge amount of memory over time if repeatedly called in a hot loop. A graphic
+// demonstrating these allocations is attached the HIVE-2399 card.
+//
+// The solution is twofold:
+// 1) store the deferred mapper on the restConfigClientGetter so that it can be reused
+// 2) cache the restConfigClientGetter for each rest config and reuse it for future reconciles
+// The type of this map is effectively map[string]*restConfigClientGetter
+var restConfigClientGetterCache sync.Map
+
+var cacheCleanMu sync.Mutex
+var lastCacheClean time.Time
+var cacheCleanInterval = time.Hour * 4
+var cacheEntryMaxAge = time.Hour * 24
+
+func init() {
+	// TODO: override cache duration variables from env
+}
+
+func RemoveRestConfigClientGetterCacheEntry(logger log.FieldLogger, key string) {
+	logger.WithField("getter-cache-key", key).Debug("removing client config cache entry")
+
+	restConfigClientGetterCache.Delete(key)
+}
+
 func (r *helper) getRESTConfigFactory(namespace string) (cmdutil.Factory, error) {
 	if r.metricsEnabled {
 		// Copy the possibly shared restConfig reference and add a metrics wrapper.
 		cfg := rest.CopyConfig(r.restConfig)
 		controllerutils.AddControllerMetricsTransportWrapper(cfg, r.controllerName, false)
 		r.restConfig = cfg
 	}
-	r.logger.WithField("cache-dir", r.cacheDir).Debug("creating cmdutil.Factory from REST client config and cache directory")
-	f := cmdutil.NewFactory(&restConfigClientGetter{restConfig: r.restConfig, cacheDir: r.cacheDir, namespace: namespace})
+
+	// If we're not using the cache, this will be the getter used. If we are using the cache,
+	// this value will be stored if the cache misses.
+	freshRestClientGetter := &restConfigClientGetter{
+		restConfig: r.restConfig,
+		cacheDir:   r.cacheDir,
+		namespace:  namespace,
+		logger:     r.logger,
+		version:    r.restConfigClientGetterVersion,
+	}
+	restClientGetter := freshRestClientGetter
+	reused := false
+
+	// HIVE-2399: reuse the same restConfigClientGetter if a cache key is provided
+	if len(r.restConfigClientGetterKey) != 0 {
+		// retrieve a cached getter, or store one if none exists for the given cache key
+		cachedRestClientGetterAny, loaded := restConfigClientGetterCache.LoadOrStore(r.restConfigClientGetterKey, freshRestClientGetter)
+
+		// if we didn't store a new getter, we check to see if we can reuse it:
+		if loaded {
+			cachedRestClientGetter := cachedRestClientGetterAny.(*restConfigClientGetter)
+
+			if cachedRestClientGetter.version == r.restConfigClientGetterVersion {
+				// if the getter has the same version, we reuse it
+				restClientGetter = cachedRestClientGetter
+				reused = true
+			} else {
+				// otherwise, update it with the fresh one
+				restConfigClientGetterCache.Store(r.restConfigClientGetterKey, freshRestClientGetter)
```

Comment:
Do these two cache ops need to be wrapped in a mutex? Or... maybe they're under the mutex [below](https://github.com/openshift/hive/pull/2246/files#diff-f52f3cee6c548137a153544d536b82bf4e67d80ad1a10546830b4433752ff038R129-R130)? /me goes hunting for call stack...